src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
+        /// Map from EcmaModule instances to field layouts within the individual modules.
         /// </summary>
-        private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+        private ModuleFieldLayoutMap _moduleFieldLayoutMap;
 
-        /// <summary>
-        /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
-        /// </summary>
-        private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
-
-        private LayoutInt _initialNonGcStaticsOffset;
+        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target)
+        {
+            _moduleFieldLayoutMap = new ModuleFieldLayoutMap(target);
+        }
 
-        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target, int numberOfTypesInModule)
+        public override ComputedStaticFieldLayout ComputeStaticFieldLayout(DefType defType, StaticLayoutKind layoutKind)
         {
-            _initialNonGcStaticsOffset = new LayoutInt(DomainLocalModuleDataBlobOffsetAsIntPtrCount * target.PointerSize + numberOfTypesInModule);
+            ComputedStaticFieldLayout layout = new ComputedStaticFieldLayout();
+            if (defType is EcmaType ecmaType)
+            {
+                // ECMA types are the only ones that can have statics
+                ModuleFieldLayout moduleFieldLayout = _moduleFieldLayoutMap.GetOrCreateValue(ecmaType.EcmaModule);
+                layout.GcStatics = moduleFieldLayout.GcStatics;
+                layout.NonGcStatics = moduleFieldLayout.NonGcStatics;
+                layout.ThreadGcStatics = moduleFieldLayout.ThreadGcStatics;
+                layout.ThreadNonGcStatics = moduleFieldLayout.ThreadNonGcStatics;
+                moduleFieldLayout.TypeToFieldMap.TryGetValue(defType, out layout.Offsets);
+            }
+            return layout;
         }
 
-        protected override void PrepareRuntimeSpecificStaticFieldLayout(TypeSystemContext context, ref ComputedStaticFieldLayout layout)
+        /// <summary>
+        /// Map from modules to their static field layouts.
+        /// </summary>
+        private class ModuleFieldLayoutMap : LockFreeReaderHashtable<EcmaModule, ModuleFieldLayout>
         {
-            layout.NonGcStatics.Size = _initialNonGcStaticsOffset;
-            layout.GcStatics.Size = LayoutInt.Zero;
-            layout.ThreadNonGcStatics.Size = LayoutInt.Zero;
-            layout.ThreadGcStatics.Size = LayoutInt.Zero;
+            /// <summary>
+            /// <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/class.h#L64">MAX_LOG2_PRIMITIVE_FIELD_SIZE</a>
+            /// </summary>
+            private const int MaxLog2PrimitiveFieldSize = 3;
+
+            /// <summary>
+            /// <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/class.h#L66">MAX_PRIMITIVE_FIELD_SIZE</a>
+            /// </summary>
+            private const int MaxPrimitiveFieldSize = 1 << MaxLog2PrimitiveFieldSize;
```

Comment:
Zach introduces a constant for this in a more central location in https://github.com/dotnet/corert/pull/6303#discussion_r216636225. Let's not introduce a local constant for this. Having multiple definitions of the same value across the codebase kind of defeats the purpose of constants like this.

Reply:
OK, it's already been merged in and I resolved the change against it.